extensions/amp-a4a/0.1/refresh-manager.js

```diff
@@ -0,0 +1,366 @@
+  }
+
+  /**
+   * Returns a function that will be invoked directly by the
+   * IntersectionObserver implementation. It will implement the core logic of
+   * the refresh lifecycle, including the transitions of the DFA.
+   *
+   * @param {!Array<!IntersectionObserverEntry>} entries
+   */
+  ioCallback_(entries) {
+    entries.forEach(entry => {
+      const refreshManagerId = entry.target.getAttribute(DATA_MANAGER_ID_NAME);
+      dev().assert(refreshManagerId);
+      const refreshManager = managers[refreshManagerId];
+      if (entry.target != refreshManager.element_) {
+        return;
+      }
+      switch (refreshManager.state_) {
+        case RefreshLifecycleState.INITIAL:
+          // First check if the element qualifies as "being on screen", i.e.,
+          // that at least a minimum threshold of pixels is on screen. If so,
+          // begin a timer, set for the duration of the minimum time on screen
+          // threshold. If this timer runs out without interruption, then all
+          // viewability conditions have been met, and we can begin the refresh
+          // timer.
+          if (entry.intersectionRatio >=
+              refreshManager.config_.visiblePercentageMin) {
+            refreshManager.state_ = RefreshLifecycleState.VIEW_PENDING;
+            refreshManager.visibilityTimeoutId_ = refreshManager.timer_.delay(
+                () => {
+                  refreshManager.state_ = RefreshLifecycleState.REFRESH_PENDING;
+                  refreshManager.startRefreshTimer_();
+                }, refreshManager.config_.continuousTimeMin);
+          }
+          break;
+        case RefreshLifecycleState.VIEW_PENDING:
+          // If the element goes off screen before the minimum on screen time
+          // duration elapses, place it back into INITIAL state.
+          if (entry.intersectionRatio <
+              refreshManager.config_.visiblePercentageMin) {
+            refreshManager.timer_.cancel(refreshManager.visibilityTimeoutId_);
+            refreshManager.visibilityTimeoutId_ = null;
+            refreshManager.state_ = RefreshLifecycleState.INITIAL;
+          }
+          break;
+        case RefreshLifecycleState.REFRESH_PENDING:
+        case RefreshLifecycleState.REFRESHED:
+        default:
+          break;
+      }
+    });
+  }
+
+  /**
+   * Initiates the refresh cycle by initiating the visibility manager on the
+   * element.
+   */
+  initiateRefreshCycle() {
+    if (!this.isRefreshable()) {
+      return;
+    }
+    switch (this.state_) {
+      case RefreshLifecycleState.INITIAL:
+        this.getIntersectionObserverWithThreshold_(
+            this.config_.visiblePercentageMin).observe(this.element_);
+        break;
+      case RefreshLifecycleState.REFRESHED:
+        this.getIntersectionObserverWithThreshold_(
+            this.config_.visiblePercentageMin).unobserve(this.element_);
+        this.getIntersectionObserverWithThreshold_(
```

Comment:
I understand that we need to `unobserve` then `observe` again in case the element is in viewport at first, but this is super confusing.
What do you think of getting rid of `REFRESHED` state. Call `unobserve()` after min duration met and call `observe()` again in `initiateRefreshCycle()` only.